tsl/src/bgw_policy/job.c

```diff
@@ -839,6 +815,233 @@ job_execute(BgwJob *job)
+	char *query;
+	List *raw_parsetree_list;
+	ListCell *lc1;
+	bool isTopLevel;
+	int commands_remaining;
+	MemoryContext parsecontext;
+	MemoryContext oldcontext;
+
+	StartTransactionCommand();
+
+	/* Build the CALL or SELECT statement depending of the "prokind" */
+	if (job->fd.config == NULL)
+		jsonstr = "NULL";
+	else
+		jsonstr = JsonbToCString(NULL, &job->fd.config->root, VARSIZE(job->fd.config));
+
+	query = psprintf("%s %s.%s(%d, %s)",
+					 (prokind == PROKIND_PROCEDURE) ? "CALL" : "SELECT",
+					 quote_identifier(NameStr(job->fd.proc_schema)),
+					 quote_identifier(NameStr(job->fd.proc_name)),
+					 job->fd.id,
+					 (job->fd.config == NULL) ? jsonstr : quote_literal_cstr(jsonstr));
+
+	debug_query_string = query;
+
+	/*
+	 * Parse the SQL string into a list of raw parse trees.
+	 *
+	 * Because we allow statements that perform internal transaction control,
+	 * we can't do this in TopTransactionContext; the parse trees might get
+	 * blown away before we're done executing them.
+	 */
+	parsecontext = AllocSetContextCreate(TopMemoryContext,
+										 "job_execute_sql parse/plan",
+										 ALLOCSET_DEFAULT_MINSIZE,
+										 ALLOCSET_DEFAULT_INITSIZE,
+										 ALLOCSET_DEFAULT_MAXSIZE);
+	oldcontext = MemoryContextSwitchTo(parsecontext);
+	raw_parsetree_list = pg_parse_query(query);
+	commands_remaining = list_length(raw_parsetree_list);
+	isTopLevel = commands_remaining == 1;
+	MemoryContextSwitchTo(oldcontext);
+
+	/*
+	 * Do parse analysis, rule rewrite, planning, and execution for each raw
+	 * parsetree.  We must fully execute each query before beginning parse
+	 * analysis on the next one, since there may be interdependencies.
+	 */
+	foreach (lc1, raw_parsetree_list)
+	{
+		RawStmt *parsetree = (RawStmt *) lfirst(lc1);
+
+#if PG_VERSION_NUM < 130000
+		const char *commandTag;
+		char completionTag[COMPLETION_TAG_BUFSIZE];
+#else
+		CommandTag commandTag;
+		QueryCompletion qc;
+#endif
+
+		List *querytree_list;
+		List *plantree_list;
+		bool snapshot_set = false;
+		Portal portal;
+		DestReceiver *receiver;
+		int16 format = 1;
+
+		/*
+		 * We don't allow transaction-control commands like COMMIT and ABORT
+		 * here.  The entire SQL statement is executed as a single transaction
+		 * which commits if no errors are encountered.
+		 */
+		if (IsA(parsetree, TransactionStmt))
+			ereport(ERROR,
+					(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+					 errmsg("transaction control statements are not allowed in jobs")));
+
+		/*
+		 * Get the command name for use in status display (it also becomes the
+		 * default completion tag, down inside PortalRun).  Set ps_status and
+		 * do any special start-of-SQL-command processing needed by the
+		 * destination.
+		 */
+		commandTag = CreateCommandTag(parsetree->stmt);
+
+		BeginCommand(commandTag, DestNone);
+
+		/* Set up a snapshot if parse analysis/planning will need one. */
+		if (analyze_requires_snapshot(parsetree))
+		{
+			PushActiveSnapshot(GetTransactionSnapshot());
```

Comment:
need a test case for this.

Reply:
You're correct... but first I'm trying a way to avoid the parser ...